File: acm_timus/1818/Main.py
import sys

class pair:
    first = 0
    second = 0
    def __init__(self):
        self.first = 0
        self.second  = 0
    def __init__(self, valFirst, valSecond):
        self.first = valFirst
        self.second = valSecond

LIMIT_SYS = 1000000

sys.setrecursionlimit(LIMIT_SYS)

def euclid(a, b):
    if(b == 0):
        return pair(1, 0)
    quotient = a // b
    remainder = a % b
    rest = euclid(b, remainder)

    x = rest.second
    y = rest.first - quotient * rest.second
    return pair(x, y)

DEBUG_MODE = bool(False)
ATTEMPT_WRITE_CORRECT = bool(True)


if DEBUG_MODE:
    fin = open("aux.in", "r")
    fout = open("aux.out", "w")
else:
    fin = sys.stdin
    fout = sys.stdout

finBuff = fin.read().split()
n = int(finBuff[0])

powNN = 1
powNN1 = 1


for index in range(1, n + 1):
    powNN = powNN * n
    powNN1 = powNN1 * (n - 1)

S = int(0)

powNAux = 1
powN1Aux = powNN1

for index in range(1, n + 1):
    S += int(finBuff[index]) * powNAux * powN1Aux
    powNAux *= n
    powN1Aux //= n - 1

A = powNN1
B = powNN

P = euclid(B, A)
B1 = P.first

sHelper = A - (S % A)

firstGuy = sHelper * B1
firstGuy %= A
firstGuy += A
firstGuy %= A

lastGuy = (firstGuy * powNN + S) // powNN1

if(lastGuy == 0):
    lastGuy = powNN

lastGuyCopy = lastGuy

if ATTEMPT_WRITE_CORRECT:

    ansString = str("")


    while (lastGuyCopy):
        ansString += str(lastGuyCopy % int(10))
        lastGuyCopy //= int(10)

    sizeString = len(ansString)
    index = sizeString - 1

    while (index >= 0):
        fout.write(ansString[index])
        index = index - 1


    fout.write('\n')

else:
    fout.write(str(lastGuy))
    fout.write('\n')

fin.close()
fout.close()

File: acm_timus/1818/Main.java
import java.util.*;
import java.io.*;
import java.math.*;
import java.math.BigInteger;


public class Main  {
    static final int maxSize = 2000 + 1;
    private static class myPair  {
        public BigInteger first;
        public BigInteger second;
        public myPair()  {
            first = BigInteger.ZERO;
            second = BigInteger.ZERO;
        }
        public myPair(BigInteger _first, BigInteger _second)  {
            first = _first;
            second = _second;
        }
    }

    private static myPair euclid(BigInteger a, BigInteger b)  {
        if(b == BigInteger.ZERO)  {
            return new myPair(BigInteger.ONE, BigInteger.ZERO);
        }
        BigInteger x, y;
        BigInteger remainder = a.mod(b);
        BigInteger quotient = a.divide(b);

        myPair rest = euclid(b, remainder);

        x = rest.second;
        y = rest.first;
        y = y.subtract(quotient.multiply(x));

        return new myPair(x, y);
    }
    public static void main(String[] args) throws FileNotFoundException, IOException {

        final var fin = new BufferedReader(new InputStreamReader(System.in));

        BigInteger n, n1;
        BigInteger[] a = new BigInteger[maxSize];
        BigInteger[] pows = new BigInteger[maxSize];
        BigInteger[] powsLess = new BigInteger[maxSize];

        pows[0] = BigInteger.ONE;
        powsLess[0] = BigInteger.ONE;

        n = nextLine(fin).nextBigInteger();
        n1 = n.subtract(BigInteger.ONE);
        int nInt = n.intValue();

        for(int i = 1;i <= nInt;i++)  {
            a[i] = fin.nextBigInteger();
        }

        for(int i = 1;i <= nInt;i++)  {
            pows[i] = pows[i - 1].multiply(n);
            powsLess[i] = powsLess[i - 1].multiply(n1);
        }
        BigInteger S = new BigInteger("0");
        for(int i = 1;i <= nInt;i++)  {
            S = S.add(a[i].multiply(pows[i - 1].multiply(powsLess[nInt + 1 - i])));
        }

        BigInteger A, B;
        A = powsLess[nInt];
        B = pows[nInt];

        myPair P = euclid(B, A);
        BigInteger B1 = P.first;

        BigInteger sHelper = S;
        sHelper = S.mod(A);
        sHelper = A.subtract(sHelper);

        BigInteger firstGuy = sHelper.multiply(B1);
        firstGuy = firstGuy.mod(A);
        firstGuy = firstGuy.add(A);
        firstGuy = firstGuy.mod(A);

        BigInteger lastGuy;

        lastGuy = firstGuy.multiply(pows[nInt]);
        lastGuy = lastGuy.add(S);
        lastGuy = lastGuy.divide(powsLess[nInt]);

        if(lastGuy == BigInteger.ZERO)
            lastGuy = pows[nInt];

        fout.println(lastGuy);

        fout.flush();

        if(DEBUG_MODE)  {
            fin.close();
            fout.close();
            ferr.close();
        }
    }
}